internals/src/compact_size.rs

```diff
@@ -0,0 +1,310 @@
+
+//! Variable length integer encoding A.K.A [`CompactSize`].
+//!
+//! An integer can be encoded depending on the represented value to save space. Variable length
+//! integers always precede an array/vector of a type of data that may vary in length.
+//!
+//! [`CompactSize`]: <https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer>
+
+use core::fmt;
+
+use crate::write_err;
+
+/// The maximum size of a serialized object in bytes or number of elements
+/// (for eg vectors) when the size is encoded as CompactSize.
+///
+/// This is `MAX_SIZE` in Bitcoin Core.
+pub const MAX_ENCODABLE_SIZE: u64 = 0x0200_0000;
+
+/// Returns the number of bytes used to encode this `CompactSize` value.
+///
+/// # Returns
+///
+/// - 1 for 0..=0xFC
+/// - 3 for 0xFD..=(2^16-1)
+/// - 5 for 0x10000..=(2^32-1)
+/// - 9 otherwise.
+#[inline]
+pub const fn encoded_size(value: u64) -> usize {
+    match value {
+        0..=0xFC => 1,
+        0xFD..=0xFFFF => 3,
+        0x10000..=0xFFFFFFFF => 5,
+        _ => 9,
+    }
+}
+
+/// Encodes `CompactSize` without allocating.
+///
+/// # Returns
+///
+/// An array with the encoded value in it as well as the length, in bytes, of the encoded value.
+///
+/// # Example
+///
+/// ```
+/// use bitcoin_internals::compact_size;
+///
+/// let (encoded, size) = compact_size::encode(32);
+/// // Typically you will want to grab the slice using `encoded[..size]`.
+/// assert_eq!(&encoded[..size], [0x20].as_slice());
+///
+/// assert_eq!(size, 1);
+/// assert_eq!(encoded, [0x20, 0, 0, 0, 0, 0, 0, 0, 0]);
+/// ```
+pub fn encode(value: u64) -> ([u8; 9], usize) {
+    let mut buf = [0_u8; 9];
+    let size;
+
+    match value {
+        0..=0xFC => {
+            size = 1;
+            buf[0] = value as u8; // Cast ok because of match.
+        }
+        0xFD..=0xFFFF => {
+            size = 3;
+            let v = value as u16; // Cast ok because of match.
+            buf[0] = 0xFD;
+            buf[1..size].copy_from_slice(&v.to_le_bytes());
+        }
+        0x10000..=0xFFFFFFFF => {
+            size = 5;
+            let v = value as u32; // Cast ok because of match.
+            buf[0] = 0xFE;
+            buf[1..size].copy_from_slice(&v.to_le_bytes());
+        }
+        _ => {
+            size = 9;
+            let v = value;
+            buf[0] = 0xFF;
+            buf[1..size].copy_from_slice(&v.to_le_bytes());
+        }
+    }
+    (buf, size)
+}
+
+/// Gets the `CompactSize` encoded at the front of `slice`.
+///
+/// Does not modify `slice`, enforces invariant that `size < CompactSize::MAX_SIZE`.
+///
+/// # Returns
+///
+/// The `CompactSize` if one exists as well as the number of bytes used to encode it.
+///
+/// # Example
+///
+/// ```
+/// use bitcoin_internals::compact_size;
+///
+/// // Extra bytes are ignored.
+/// let (compact, _) = compact_size::decode(&[0x20, 0xAB, 0xCD]).unwrap();
+/// assert_eq!(compact, 32);
+/// ```
+pub fn decode(slice: &[u8]) -> Result<(u64, usize), DecodeError> {
```

Comment:
Rather than returning `usize` and forcing the callers to recompute it's better to take slice as `&mut &[u8]` and mutate it to update the start of the slice. You can then avoid a bunch of redundant code.

Reply:
I'll take a look, thanks.

Reply:
With fresh eyes I still don't understand what you mean, can you extrapolate please?

Reply:
This signature is better:

`pub fn decode(slice: &mut &[u8]) -> Result<u64, DecodeError>`

Where the function updates the slice by moving start by the number of consumed bytes.

Because callers always want to update their positions in the slice, this method is more helpful and doing it manually is annoying and error-prone.

Reply:
I thought that was what you might be meaning but in both callsites in this PR the caller does not update the slice.

Reply:
It does, check again.

Reply:
Oh you are right, the iterator does update it. The other callsite (in `element_at`), does not and does not have a mutable reference to it ether (called from `nth`).  Also `decode_cursor` does not mutate the slice, I still rekon we should leave it as is. Thougts?

Reply:
All of them mutate the slice in the sense that they re-construct the new one in the exact same way - to start after the compact size bytes. They could as well make a `mut bytes` copy and mutate it. You could argue we should return `(u64, &[u8])` to keep the style functional which is not terrible either, I'm just used to mutating style because of deserializers. Also I'm not sure what the borrow checker will think of it although it's likely fine.